### PR TITLE
add patch_inplace function

### PR DESCRIPTION
**Pull request by wolfv**
I am trying to `patch` a >100Mb JSON file and it turns out that the copying / destruction of JSON objects takes quite long.

Not allocating a new JSON for each patch helps a lot in terms of speed (>100x improvement).

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for these kind of bugs). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](https://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.

### CODE REVIEW

**Comment by falbrechtskirchinger** [comment]
A couple of things:
* Please run `make amalgamate` (see [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change)).
* I think `patch_inplace` would be the better name, so it appears next to `patch` when sorted.
* The documentation needs to be updated. I suggest you move `docs/mkdocs/docs/api/basic_json/patch.md` to `patch_inplace.md` and add a new `patch.md`, explaining `patch()` in terms of `patch_inplace()` and linking to `patch_inplace.md` for details. Example files need to be duplicated and adjusted.

(Separately, we should address the quite frequent failure of MinGW in CI. Can we rehost the MinGW package without violating the license?)

**Comment by wolfv** [comment]
@falbrechtskirchinger thanks for the notes. What do you think about keeping docs in a single document? Otherwise there would be quite some duplication.

This really drastically improves the performance for our use case as copying and destroying a 100Mb json file is _really_ slow (on the tune of ~1 second per patch, vs a couple nanoseconds without the copy).

I wonder if the destruction could be optimized in #3583

**Comment by falbrechtskirchinger** [comment]
> @falbrechtskirchinger thanks for the notes. What do you think about keeping docs in a single document? Otherwise there would be quite some duplication.

@wolfv I don't think there's precedent for similarly named functions being documented in one file and while I agree with you about the duplication, @nlohmann will probably want a dedicated `patch_inplace.md` file. Let's wait and see.

You're at least missing an entry in `docs/mkdocs/mkdocs.yml`.
```yaml
- 'patch_inplace': api/basic_json/patch.md
```

Also, can you please prepare the examples in `docs/examples` and include them once a decision has been made as to where `patch_inplace()` should be documented?
Basically, copy `patch.cpp` -> `patch_inplace.cpp` and `patch.output` -> `patch_inplace.output` and modify `patch_inplace.cpp`:

```cpp
    // output original document
    std::cout << std::setw(4) << doc << "\n\n";

    // apply the patch inplace
    doc.patch_inplace(patch);

    // output patched document
    std::court << std::setw(4) << doc << std::endl;
```

Did you run `make amalgamate`? You need to commit the generated `single_include/nlohmann/json.hpp` file as well.

**Comment by nlohmann** [comment]
> @wolfv I don't think there's precedent for similarly named functions being documented in one file and while I agree with you about the duplication, @nlohmann will probably want a dedicated `patch_inplace.md` file. Let's wait and see.

Yes, please do so.

**Review by nlohmann — CHANGES_REQUESTED** [review]

**Comment by wolfv** [comment]
Thanks for the review! I applied your suggestions.

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Review by falbrechtskirchinger — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Note to me:

- [ ] Add new documentation to mkdocs.

**Comment by wolfv** [comment]
@nlohmann thanks for approving! Is the remaining task something I should do?

**Comment by nlohmann** [comment]
Thanks for your work!

> @nlohmann thanks for approving! Is the remaining task something I should do?

It's basically checking that the new Markdown file is added to docs/mkdocs/docs/api/basic_json/index.md, docs/mkdocs/mkdocs.yml, and docs/docset/docSet.sql. I'll take care of it.
